New editor - Save subtitles in Opencast

Arnei opened the pull request:
A new feature is currently in development for the new editor -  a subtitle editor. It will allow editing or creating subtitle tracks in the new editor UI and then saving them in Opencast. This modifies the editor service endpoint to allow the saving of subtitles. 

The POST "{mediaPackageId}/edit.json" endpoint is extended to accept a list of subtitles and flavors which are to be saved in the mediapackage. Subtitles in existing flavors are overwritten, while subtitles in new flavors are added as new tracks.

Does not perform any checks (e.g. validity) on the submitted subtitles.

Depends on #3601

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

Arnei committed:
Added subtitle saving to the editor endpoint

The new editor can send subtitle objects for saving. This commit adds or overwrites existing subtitles to the corresponding mediapackage

Arnei committed:
Added nullcheck

Arnei committed:
Fixed type in EditorServiceRemoteImpl

metzenseifner commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -420,6 +422,84 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
+   */
+  private void addSubtitleTrack(MediaPackage mediaPackage, List<PostEditingData.Subtitle> subtitles)
+          throws IOException {
+    if (subtitles != null) {
```
Consider replacing with `Objects.requireNonNull(subtitles, "Argument subtitles may not be null");`, thereby throwing a meaningful NullPointerException. Reasoning: subtitles being null is clearly an invalid state. As it is now, the caller has no way of knowing that something went awry.

Arnei replied:
Thanks, will do.

metzenseifner commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -786,6 +867,12 @@ public void setEditData(String mediaPackageId, EditingData editingData) throws E
 
+    try {
+      addSubtitleTrack(mediaPackage, editingData.getSubtitles());
+    } catch (IOException e) {
```
} catch (IOException | NullPointerException e) {

(Should you decided to heed my suggestion about throwing NullPointerException)

Arnei replied:
Caught the NullPointerException in a different clause in [7a78ba0](https://github.com/opencast/opencast/pull/3604/commits/7a78ba0efafe259f6663aaf6efaf818e59d61f00), because it is perfectly valid for the frontend to not send any subtitles.

metzenseifner reviewed (changes requested):
I think my suggestions would help give the user better feedback in the problem case.

lkiesow commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -420,6 +422,84 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
+      // Put updated filename in working file repository and update the track.
+      InputStream is = IOUtils.toInputStream(subtitle.getSubtitle(), "UTF-8");
+      URI subtitleUri = workspace.put(mediaPackage.getIdentifier().toString(), subtitleId, "subtitle.vtt",
+              //                  EditorService.TARGET_FILE_NAME,
```
```suggestion
```

lkiesow commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -420,6 +422,84 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
+      String trackId = null;
+
+      // Check if subtitle already exists
+      for (Track t : mediaPackage.getTracks()) {
```
This allows users to easily replace any track with any content. It may be a good idea to limit this to caption flavors only.

Arnei replied:
Yeah, I wasn't sure what the bigger issue was, being able to replace any track, or being limited to the (implicit standard)`captions/*` flavor.

lkiesow replied:
This is now caught by
```java
if (!subtitle.getFlavor().matches(captionsFlavor)) {
```

lkiesow commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -420,6 +422,84 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
+      }
+
+      // Put updated filename in working file repository and update the track.
+      InputStream is = IOUtils.toInputStream(subtitle.getSubtitle(), "UTF-8");
```
It's a good idea to always close input streams

```suggestion
      try (InputStream is = IOUtils.toInputStream(subtitle.getSubtitle(), "UTF-8")) {
```

lkiesow commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -420,6 +422,84 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
+      }
+      track.setChecksum(null);
+
+      if (oldTrackURI != null) {
```
The old and new URIs could be identical, in which case you would not want to delete this here

lkiesow commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -420,6 +422,84 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
+      } catch (AssetManagerException e) {
+        logger.error("Error while adding the updated media package ({}) to the archive", mediaPackage.getIdentifier(),
+                e);
+        throw new IOException(e);
```
Don't log an error if you return an error. Add the message to the thrown exception:

```suggestion
        throw new IOException("Error while adding the updated media package " + mediaPackage + " to the archive", e);
```

lkiesow commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -420,6 +422,84 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
+                e);
+        throw new IOException(e);
+      }
+    }
```
You leave files in the workspace. Once added to the asset manager, you should call `workspace.delete(…)` to remove the subtitle from there.

lkiesow commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -420,6 +422,84 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
+      track.setIdentifier(subtitleId);
+      for (String tag : getSmilCatalogTags()) {
+        track.addTag(tag);
+      }
```
This does not fit the documentation for these tags which is:

```properties
# Sets the tag that should be applied to the generated smil catalog                                                      
# Default: archive                                                                                                       
#smil.catalog.tags=archive 
```

A subtitle is no SMIL catalog. I would just drop these lines:

```suggestion
```

lkiesow reviewed (changes requested):
I found a few minor problems, but nothing major. Before we merge this, we may want to sit down and discuss how subtitles should actually be named/tagged/flavored though. This code (and related pull requests) feel like we have too many implicit assumptions about how things could work.

lkiesow commented:
Maybe you missed the CI failure, but this does not build any longer

lkiesow commented on modules/editor-service-api/src/main/java/org/opencastproject/editor/api/EditingData.java:
```diff
@@ -61,10 +61,10 @@ public EditingData(List<SegmentData> segments, List<TrackData> tracks, List<Work
   }
 
-  public static EditingData parse(String json) {
+  public static PostEditingData parse(String json) {
```
Why would a static method of the class `EditingData` return `PostEditingData`?

lkiesow reviewed (changes requested):
I'll write it down here as well as on the frontend pull request since it touches both parts:

I did only briefly skim through the code, but if I interpret the code correctly, subtitles are loaded from the internal publication, but then saved to the asset manager. This means that you don't necessary get what you save the next time you open the editor. It also means that you may overwrite completely different subtitle files and you might also miss subtitle files which are only in the asset manager.

My suggestion would be to always get the subtitles from the asset manager?

Arnei commented:
Rebased this on r/12.x

Arnei committed:
Make requested changes again

Some of the requested changes got lost during rebase. Hopefully this commit solves the requests again.

Arnei committed:
Get subtitles from asset manager

lkiesow commented on modules/editor-service-api/src/main/java/org/opencastproject/editor/api/PostEditingData.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Wrapper for additional data that may be received by a post request from the editor
+ */
+public final class PostEditingData extends EditingData {
```
Does it make sense to use this for both ways from and to the editor? The only difference is the subtitle and we could provide it using this method as well.

_(see also comment above)_

lkiesow commented on modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java:
```diff
@@ -437,6 +447,82 @@ void addSmilToArchive(MediaPackage mediaPackage, final Smil smil) throws IOExcep
+    for (PostEditingData.Subtitle subtitle : subtitles) {
+      if (!subtitle.getFlavor().matches(captionsFlavor)) {
+        break;
+      }
```
Maybe raise an exception which ends in a 400 Bad Request for the user, so it is clear that the request wasn't processed successfully and, in particular, wasn't saved as sent by the user. Also, maybe not save anything if half of the request is bad? Otherwise, you don't know the state you end up with.

lkiesow reviewed (changes requested):
Generally looks good and you would now also get the saved subtitles if access control lets you, which might be a problem in a cluster with the current solution (see below). 

We also found another minor issue which should never be a problem if everything goes well, but still might be worth fixing.

For getting the subtitle to the frontend, right now, you just replace the internal publication with a track containing a direct URI to the asset manager. I'm not sure if you can always access the asset manager directly like this, e.g., if you are on a presentation node.

You could use `EditingData` for sending the subtitle instead of replacing parts of the internal publications. This also has the advantage that the codes becomes simpler since you don't need `PostEdittingData` and `edit.json` has always the same structure.

Arnei committed:
Send subtitle separately from tracks

Allows us to remove PostEditingData.java in favor of a consolidated EditingData.java

Arnei committed:
In case of subtitle flavor mismatching, fail more explicitely

Arnei committed:
Send subtitles as blob

Instead of sending an URI to the asset manager (which can potentially cause access issues on certain distributed setups), send the whole text as a giant string instead.

Arnei committed:
Merge remote-tracking branch 'upstream/r/12.x' into new-editor-save-subtitles

# Conflicts:
#	modules/editor-service/src/main/java/org/opencastproject/editor/EditorServiceImpl.java

Arnei committed:
Don't save any subtitles if the request is faulty somehow

Arnei committed:
Fixed faulty import

Arnei commented:
> Generally looks good and you would now also get the saved subtitles if access control lets you, which might be a problem in a cluster with the current solution (see below).
> 
> We also found another minor issue which should never be a problem if everything goes well, but still might be worth fixing.
> 
> For getting the subtitle to the frontend, right now, you just replace the internal publication with a track containing a direct URI to the asset manager. I'm not sure if you can always access the asset manager directly like this, e.g., if you are on a presentation node.
> 
> You could use `EditingData` for sending the subtitle instead of replacing parts of the internal publications. This also has the advantage that the codes becomes simpler since you don't need `PostEdittingData` and `edit.json` has always the same structure.

To avoid any problems with accessing the asset manager, I've changed it just send the complete subtitle file contents instead. As you already mentioned, we already do this from Frontend-to-Backend anyway, so now it's at least consistent for both ways. Which also allows for removal of `PostEditingData`

Arnei committed:
Don't actually remove subtitle tracks from tracklist

There's no point to it anymore and it will cause an exception if trackList is fixedSize for whatever reason.

lkiesow reviewed (approved):
Looks good and seems to work as expected. It should also be fully backwards compatible. Merging.
